kylin-spark-project/kylin-spark-common/src/main/spark31/org/apache/spark/sql/execution/KylinJoinSelection.scala

```diff
@@ -0,0 +1,285 @@
+    plan.stats.sizeInBytes < conf.autoBroadcastJoinThreshold * conf.numShufflePartitions
+  }
+
+  /**
+   * Returns whether plan a is much smaller (3X) than plan b.
+   *
+   * The cost to build hash map is higher than sorting, we should only build hash map on a table
+   * that is much smaller than other one. Since we does not have the statistic for number of rows,
+   * use the size of bytes here as estimation.
+   */
+  private def muchSmaller(a: LogicalPlan, b: LogicalPlan): Boolean = {
+    a.stats.sizeInBytes * 3 <= b.stats.sizeInBytes
+  }
+
+  private def canBuildRight(joinType: JoinType): Boolean = joinType match {
+    case _: InnerLike | LeftOuter | LeftSemi | LeftAnti | _: ExistenceJoin => true
+    case _ => false
+  }
+
+  private def canBuildLeft(joinType: JoinType): Boolean = joinType match {
+    case _: InnerLike | RightOuter => true
+    case _ => false
+  }
+
+  private def broadcastSide(
+                             canBuildLeft: Boolean,
+                             canBuildRight: Boolean,
+                             left: LogicalPlan,
+                             right: LogicalPlan): BuildSide = {
+
+    def smallerSide =
+      if (right.stats.sizeInBytes <= left.stats.sizeInBytes) BuildRight else BuildLeft
+
+    if (canBuildRight && canBuildLeft) {
+      // Broadcast smaller side base on its estimated physical size
+      // if both sides have broadcast hint
+      smallerSide
+    } else if (canBuildRight) {
+      BuildRight
+    } else if (canBuildLeft) {
+      BuildLeft
+    } else {
+      // for the last default broadcast nested loop join
+      smallerSide
+    }
+  }
+
+  private def canBroadcastByHints(joinType: JoinType, left: LogicalPlan, right: LogicalPlan, hint: JoinHint)
+  : Boolean = {
+    val buildLeft = canBuildLeft(joinType) && hintBroadcast(hint.leftHint)
+    val buildRight = canBuildRight(joinType) && hintBroadcast(hint.rightHint)
+    buildLeft || buildRight
+  }
+
+  private def broadcastSideByHints(joinType: JoinType, left: LogicalPlan, right: LogicalPlan, hint: JoinHint)
+  : BuildSide = {
+    val buildLeft = canBuildLeft(joinType) && hintBroadcast(hint.leftHint)
+    val buildRight = canBuildRight(joinType) && hintBroadcast(hint.rightHint)
+    broadcastSide(buildLeft, buildRight, left, right)
+  }
+
+  private def canBroadcastBySizes(joinType: JoinType, left: LogicalPlan, right: LogicalPlan)
+  : Boolean = {
+    val buildLeft = canBuildLeft(joinType) && canBroadcast(left)
+    val buildRight = canBuildRight(joinType) && canBroadcast(right)
+    buildLeft || buildRight
+  }
+
+  private def broadcastSideBySizes(joinType: JoinType, left: LogicalPlan, right: LogicalPlan)
+  : BuildSide = {
+    val buildLeft = canBuildLeft(joinType) && canBroadcast(left)
+    val buildRight = canBuildRight(joinType) && canBroadcast(right)
+    broadcastSide(buildLeft, buildRight, left, right)
+  }
+
+    def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
```

Comment:
It's better to implementation this strategy according to the `JoinSelection` in Spark 3.1, the implementation in Spark 3.1 is very different from the one in Spark 2.4.

Reply:
@zzcclp Hi, I've done the rewrite, but a little different from the 2.4 implementation. I've found in 2.4 implementation that it acquires memory twice for both left and right side. For example:

```
case ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, condition, left, right)
  if canBroadcastBySizes(joinType, left, right) =>
  val buildSide = broadcastSideBySizes(joinType, left, right)
  Seq(joins.BroadcastHashJoinExec(
    leftKeys, rightKeys, joinType, buildSide, condition, planLater(left), planLater(right)))
```

`canBroadcastBySizes` and `broadcastSideBySizes` both tries to acquire memory for two sides. Is that a mistake? Should we add a method like `tryAcquireMemory` in `JoinMemoryManager` for checks and acquire memory only if we decide to create BHJ?

Reply:
It seems a mistake. Thanks for your reporting about this issue, I will check this later.